packages/opencensus-web-propagation-tracecontext/src/tracecontext-format.ts

```diff
@@ -0,0 +1,117 @@
+/** The tracestate header key. */
+const TRACE_STATE = 'tracestate';
+/** Regular expression that represents a valid traceparent header. */
+const TRACE_PARENT_REGEX = /^[\da-f]{2}-[\da-f]{32}-[\da-f]{16}-[\da-f]{2}$/;
+
+/**
+ * Parses a traceparent header value into a SpanContext object, or null if the
+ * traceparent value is invalid.
+ */
+export function traceParentToSpanContext(traceParent: string): SpanContext|
+    null {
+  const match = traceParent.match(TRACE_PARENT_REGEX);
+  if (!match) return null;
+  const parts = traceParent.split('-');
+  const traceId = parts[1];
+  if (traceId.match(/^0*$/)) return null;  // All zeros trace ID is invalid.
+  const spanId = parts[2];
+  // tslint:disable-next-line:ban Needed to parse hexadecimal.
+  const options = parseInt(parts[3], 16);
+  return {traceId, spanId, options};
+}
+
+/**
+ * Converts a SpanContext object to a traceparent header value. This assumes
+ * that the length of the `traceId`, `spanId` and `options` are all valid.
+ */
+export function spanContextToTraceParent(spanContext: SpanContext): string {
+  return `00-${spanContext.traceId}-${spanContext.spanId}-0${
+      (spanContext.options || DEFAULT_OPTIONS).toString(16)}`;
+}
+
+/**
+ * Propagates span context through Trace Context format propagation.
+ *
+ * Based on the Trace Context specification:
+ * https://w3c.github.io/distributed-tracing/report-trace-context.html
+ */
+export class TraceContextFormat implements Propagation {
+  /**
+   * Gets the trace context from a request headers. If there is no trace context
+   * in the headers, or if the parsed `traceId` or `spanId` is invalid, an empty
+   * context is returned.
+   */
+  extract(getter: HeaderGetter): SpanContext|null {
+    const traceParentHeader = getter.getHeader(TRACE_PARENT);
+    if (!traceParentHeader) return null;
+    const traceParent = Array.isArray(traceParentHeader) ?
+        traceParentHeader[0] :
+        traceParentHeader;
+    const spanContext = traceParentToSpanContext(traceParent);
+    if (!spanContext) return null;
+
+    const traceStateHeader = getter.getHeader(TRACE_STATE);
+    if (traceStateHeader) {
+      spanContext.traceState = Array.isArray(traceStateHeader) ?
+          traceStateHeader.join(',') :
+          traceStateHeader;
+    }
+    return spanContext;
+  }
+
+  /**
+   * Adds a trace context `traceparent` and `tracestate` (if set in
+   * `spanContext`) to the request headers.
+   */
+  inject(setter: HeaderSetter, spanContext: SpanContext): void {
+    setter.setHeader(TRACE_PARENT, spanContextToTraceParent(spanContext));
+    if (spanContext.traceState) {
+      setter.setHeader(TRACE_STATE, spanContext.traceState);
+    }
+  }
+
+  /**
+   * Generate SpanContext.
+   *
+   * Context parts are based on section 2.2.2 of TraceContext spec.
+   */
+  generate(): SpanContext {
+    return {
+      traceId: randomTraceId(),
+      spanId: randomSpanId(),
+      options: DEFAULT_OPTIONS,
+      traceState: undefined,
```

Comment:
This(`traceState`) is an optional param, I think it is safe to remove instead of assigning `undefined`. I can make similar change in OC Node repo.